File: src/shared/golang/backend/engagements/engagements.go
package engagements

import (
	"github.com/jmoiron/sqlx"
	"gitlab.com/grchive/grchive-v2/shared/backend/roles"
	"time"
)

type Engagement struct {
	Id          int64          `db:"id"`
	Name        string         `db:"name"`
	Description string         `db:"description"`
	OrgId       int64          `db:"org_id"`
	CreatedTime time.Time      `db:"created_time"`
	StartTime   *time.Time     `db:"start_time"`
	EndTime     *time.Time     `db:"end_time"`
	IsClosed    bool           `db:"is_closed"`
	Roles       *[]*roles.Role `db:"roles"`
}

type EngagementManager struct {
	db *sqlx.DB
}

func CreateEngagementManager(db *sqlx.DB) *EngagementManager {
	return &EngagementManager{
		db: db,
	}
}

File: src/webapp/client/src/ts/types/engagements.ts
import formatRFC3339 from 'date-fns/formatRFC3339'
import { Role } from '@client/ts/types/roles'

export interface RawEngagement {
    Id            : number
    Name          : string
    Description   : string
    OrgId         : number
    CreatedTime   : Date
    StartTime     : Date | null
    EndTime       : Date | null
    IsClosed      : boolean
    Roles         : Role[] | null
}

export interface RawEngagementScopingStats {
    NumRisks :                          number
    NumControls:                        number
    NumGLAccounts:                      number
    NumFinanciallyRelevantGLAccounts:   number
    NumVendors:                         number
    NumVendorProducts:                  number
}

export function createEmptyEngagement() : RawEngagement {
    return {
        Id: -1,
        Name: '',
        Description: '',
        OrgId: -1,
        CreatedTime: new Date(),
        StartTime: null,
        EndTime: null,
        IsClosed: false,
        Roles: null,
    }
}

export function cleanRawEngagementFromJSON(r : RawEngagement) {
    r.CreatedTime = new Date(r.CreatedTime)
    if (!!r.StartTime) {
        r.StartTime = new Date(r.StartTime)
    }

    if (!!r.EndTime) {
        r.EndTime = new Date(r.EndTime)
    }
}

export function engagementToJson(r : RawEngagement) : any {
    let ret : any = {
        ...r
    }

    ret.CreatedTime = formatRFC3339(r.CreatedTime)

    if (!!r.StartTime) {
        ret.StartTime = formatRFC3339(r.StartTime)
    }

    if (!!r.EndTime) {
        ret.EndTime = formatRFC3339(r.EndTime)
    }

    return ret
}
